### PR TITLE
FIX-602: reverse

### PR DESCRIPTION
**Pull request by DenisYaroshevskiy**
This is a stub to do the reverse first issue

### CODE REVIEW

File: include/eve/detail/function/simd/x86/swizzle.hpp [hunk 1]
```diff
@@ -190,7 +190,7 @@ namespace eve::detail
           if constexpr(match(c,category::int64x4 ,category::uint64x4) )
           {
             auto vc = bit_cast(v, as_<as_floating_point_t<that_t>>{});
-            return bit_cast( v[q] , as(s));
+            return bit_cast( basic_swizzle(vc, q) , as(s));
```
Comment (jfalcou) [inline include/eve/detail/function/simd/x86/swizzle.hpp:193]: if you do this, then the swizzle won't be optimized by the find_optimized_pattern.
Is there a reason you changed that ?

File: include/eve/detail/function/simd/arm/neon/swap_adjacent_groups.hpp [hunk 2]
```diff
@@ -53,7 +53,12 @@ namespace eve::detail
         else  if constexpr( c == category::float64x2  )
         {
           if constexpr(current_api > asimd) return vextq_f64(v,v,1);
-          else                              return that_t(v.get(1),v.get(0));
+          else
+          {
+            auto u64 = vreinterpretq_u64_f64(v);
```
Comment (jfalcou) [inline include/eve/detail/function/simd/arm/neon/swap_adjacent_groups.hpp:58]: does it gives better codegen than before ?

File: include/eve/detail/function/simd/arm/neon/reverse.hpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  template<typename T, typename N, arm_abi ABI>
+  EVE_FORCEINLINE wide<T,N,ABI> reverse_( EVE_SUPPORTS(neon128_), wide<T,N,ABI> v) noexcept
+  {
+    constexpr bool one_instruction_basic_swizzle = (current_api >= asimd) || std::same_as<ABI, arm_64_>;
+
+    // two reverse instructions seem better than one table lookup.
+    if constexpr ( N() >= 8 && one_instruction_basic_swizzle) return reverse_(EVE_RETARGET(cpu_), v);
+    else
+    {
+      if constexpr ( N() >= 16 ) v = eve::swap_adjacent_groups(v, lane<8>);
```
Comment (jfalcou) [inline include/eve/detail/function/simd/arm/neon/reverse.hpp:24]: can you clarify the algorithm here ? Why is it enough ?

File: include/eve/detail/function/reverse.hpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+//==================================================================================================
+#pragma once
+
+#include <eve/detail/overload.hpp>
+#include <array>
+
+namespace eve
+{
+  EVE_MAKE_CALLABLE(reverse_, reverse);
+
```
Comment (jfalcou) [inline include/eve/detail/function/reverse.hpp:16]: We usually provide a pattern generator too.

**Review by jfalcou — CHANGES_REQUESTED** [review]

File: include/eve/detail/function/simd/x86/swizzle.hpp (hunk 1 above)
Comment (DenisYaroshevskiy) [reply include/eve/detail/function/simd/x86/swizzle.hpp:193]: that was the point. 

We don't call main swizzle from the parts. Specifically that caused me a recursion when I fall back on basic swizzle.

File: include/eve/detail/function/simd/arm/neon/swap_adjacent_groups.hpp (hunk 2 above)
Comment (DenisYaroshevskiy) [reply include/eve/detail/function/simd/arm/neon/swap_adjacent_groups.hpp:58]: At least it should. I mean it's one instruction.

File: include/eve/detail/function/reverse.hpp (hunk 4 above)
Comment (DenisYaroshevskiy) [reply include/eve/detail/function/reverse.hpp:16]: Oh - I saw that. What's the point?

File: include/eve/detail/function/simd/arm/neon/swap_adjacent_groups.hpp (hunk 2 above)
Comment (DenisYaroshevskiy) [reply include/eve/detail/function/simd/arm/neon/swap_adjacent_groups.hpp:58]: Oh...
So there is smth nuts going on:
https://godbolt.org/z/vqnr93fh6

We generate some ultra hot garbage. I have no idea what's up.

And apparently these casts should only be avaliable on aarch64,  but our gcc arm builds them fine. What's up with that?
https://godbolt.org/z/vd6M86Px3

File: include/eve/detail/function/simd/arm/neon/reverse.hpp (hunk 3 above)
Comment (DenisYaroshevskiy) [reply include/eve/detail/function/simd/arm/neon/reverse.hpp:24]: Well, you can't really reverse 16 chars with one instruction.
But you can:
Swap two 64 bit groups => reverse each halve (so - swap 32 bit groups => reverse each halve etc)

Comment (jfalcou) [reply include/eve/detail/function/simd/arm/neon/reverse.hpp:24]: ok

File: include/eve/detail/function/reverse.hpp (hunk 4 above)
Comment (jfalcou) [reply include/eve/detail/function/reverse.hpp:16]: so we have a shortcut for user `v[ reverse_n<8> ]`

File: include/eve/detail/function/simd/x86/swizzle.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/detail/function/simd/x86/swizzle.hpp:193]: but then if this casted to float can be optimized, it wont ??

Comment (DenisYaroshevskiy) [reply include/eve/detail/function/simd/x86/swizzle.hpp:193]: If we are here, it's already couldn't be optimized.

File: include/eve/detail/function/reverse.hpp (hunk 4 above)
Comment (DenisYaroshevskiy) [reply include/eve/detail/function/reverse.hpp:16]: Discussed. I think we shoul delete it everywhere.

**Review by jfalcou — APPROVED** [review]

**Pull request merged**